torch/nn/parallel/distributed.py

```diff
@@ -654,6 +657,30 @@ def _register_comm_hook(self, state: object, hook: callable):
             >>>     fut.set_result(bucket.get_tensors())
             >>>     return fut
 
+        Example::
+            Below is an example of a simple allreduce hook.
+
+            >>> ddp._register_comm_hook(state = None, hook = allreduce)
+
+            >>> def allreduce(state: object, bucket: dist.GradBucket): -> torch.futures.Future
+            >>>     work = dist.allreduce(bucket.get_tensors())
+            >>>     return work.get_future()
+
+        Example::
+            Below is an example of a Parallel SGD algorithm where gradients are encoded before
+            allreduce, and then decoded after allreduce.
+
+            >>> ddp._register_comm_hook(state = None, hook = encode_and_decode)
+
+            >>> def encode_and_decode(state: object, bucket: dist.GradBucket): -> torch.futures.Future
```

Comment:
`GradBucket` is not in master do, and it's also not available at `c10d/init.cpp`. Could you please add docs for that? And since `_register_comm_hook` is a private API, I guess we need to make `GradBucket` and `get_tensors` private as well?

Reply:
@pritamdamania87 I believe I asked a similar question before. Could you comment on this?

Reply:
GradBucket is actually in c10d/init.cpp https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L139, we can probably just add some docs there.

Agree on making `GradBucket` private, I guess `get_tensors` would be private since the class `GradBucket` is private as well.